docs/tutorials/auth.md

```diff
@@ -245,3 +245,155 @@ Access to path `/` would be granted to everyone in `example.org`, however path `
 ```
 oauthOidcAnyClaims(...) -> oidcClaimsQuery("/login:groups.#[==\"appX-Tester\"]", "/:@_:email%\"*@example.org\"")
 ```
+
+## OAuth2 authorization grant flow
+
+[Authorization grant flow](https://tools.ietf.org/html/rfc6749#section-1.3) is a mechanism
+to coordinate between a user-agent, a client, and an authorization server to obtain an OAuth2 
+access token for a user. Skipper supports the flow with the `oauthGrant()` filter. 
+It works as follows:
+
+1. A user makes a request to a route with `oauthGrant()`.
+1. The filter checks whether the request has a cookie called `oauth-grant`. If it does not, or
+   if the cookie and its tokens are invalid, it redirects the user to the OAuth2 provider's 
+   authorization endpoint.
+1. The user logs into the external OAuth2 provider, e.g. by providing a username and password.
+1. The provider redirects the user back to Skipper with an authorization code, using the 
+   callback URL parameter which was part of the previous redirect. The callback route must
+   have a `grantCallback()` filter defined.
+1. Skipper calls the provider's token URL with the authorization code, and receives a response 
+   with the access and refresh tokens.
+1. Skipper stores the tokens in an `oauth-grant` cookie which is stored in the user's browser.
+1. Subsequent calls to any route with an `oauthGrant()` filter will now pass as long as the
+   access token is valid.
+
+### Encrypted cookie tokens
+
+The cookie set by the `oauthGrant()` filter contains the OAuth2 access and refresh tokens in 
+encrypted form. This means Skipper does not need to persist any session information about users, 
+while also not exposing the tokens to users.
+
+### Token refresh
+
+The `oauthGrant()` filter also supports token refreshing. Once the access token expires and
+the user makes another request, the filter automatically refreshes the token and sets the
+updated cookie in the response.
+
+### Instructions
+
+To use authorization grant flow, you need to:
+
+1. Configure OAuth2 credentials.
```

Comment:
Users should know the valid callback URL we automatically create, before they create credentials, because providers will ask for the callback URL, and if they do not know before, then they have to delete and recreate the credential.